File: lib/excelGenerator.js
/**
 * excelGenerator.js — Professional Excel (.xlsx) Workbook Generator
 *
 * Generates genuine Microsoft Excel (.xlsx) workbooks with:
 *  - Formatted header row (Royal Blue fill #2563EB, Bold White text, centered)
 *  - Alternating zebra row striping (#F8FAFC / #FFFFFF)
 *  - Number / currency / percentage auto-detection
 *  - Auto-fitted column widths with safety padding
 *  - Thin crisp grid cell borders
 *
 * Dynamic lazy-loading for fast page compilation & light bundle.
 */

export async function generateExcelWorkbook({ headers = [], rows = [], sheetName = 'Table Data' }) {
  const mod = await import('exceljs/dist/exceljs.min.js');
  const ExcelJS = mod.default || mod;

  const workbook = new ExcelJS.Workbook();
  workbook.creator = 'BestAIToolsFree Image to Excel';
  workbook.lastModifiedBy = 'BestAIToolsFree';
  workbook.created = new Date();
  workbook.modified = new Date();

  const worksheet = workbook.addWorksheet(sheetName, {
    views: [{ showGridLines: true }],
  });

  // 1. Add Headers
  if (headers && headers.length > 0) {
    const headerRow = worksheet.addRow(headers);
    headerRow.height = 28;

    headerRow.eachCell((cell) => {
      cell.fill = {
        type: 'pattern',
        pattern: 'solid',
        fgColor: { argb: 'FF2563EB' },
      };
      cell.font = {
        name: 'Segoe UI',
        size: 11,
        bold: true,
        color: { argb: 'FFFFFFFF' },
      };
      cell.alignment = {
        vertical: 'middle',
        horizontal: 'center',
        wrapText: true,
      };
      cell.border = {
        top: { style: 'thin', color: { argb: 'FF1D4ED8' } },
        left: { style: 'thin', color: { argb: 'FF1D4ED8' } },
        bottom: { style: 'medium', color: { argb: 'FF1E40AF' } },
        right: { style: 'thin', color: { argb: 'FF1D4ED8' } },
      };
    });
  }

  // 2. Add Data Rows
  if (rows && rows.length > 0) {
    rows.forEach((rowData, rowIdx) => {
      const typedRow = rowData.map((val) => {
        if (typeof val !== 'string') return val ?? '';
        const trimmed = val.trim();
        if (!trimmed) return '';

        const cleanNum = trimmed.replace(/[\$,]/g, '');
        if (/^-?\d+(\.\d+)?$/.test(cleanNum) && !isNaN(Number(cleanNum))) {
          return Number(cleanNum);
        }
        return trimmed;
      });

      const row = worksheet.addRow(typedRow);
      row.height = 22;
      const isEven = rowIdx % 2 === 0;

      row.eachCell((cell) => {
        cell.font = {
          name: 'Segoe UI',
          size: 10.5,
          color: { argb: 'FF1E293B' },
        };

        cell.fill = {
          type: 'pattern',
          pattern: 'solid',
          fgColor: { argb: isEven ? 'FFFFFFFF' : 'FFF8FAFC' },
        };

        const isNumber = typeof cell.value === 'number';
        cell.alignment = {
          vertical: 'middle',
          horizontal: isNumber ? 'right' : 'left',
          wrapText: false,
        };

        cell.border = {
          top: { style: 'thin', color: { argb: 'FFE2E8F0' } },
          left: { style: 'thin', color: { argb: 'FFE2E8F0' } },
          bottom: { style: 'thin', color: { argb: 'FFE2E8F0' } },
          right: { style: 'thin', color: { argb: 'FFE2E8F0' } },
        };
      });
    });
  }

  // 3. Auto-fit Column Widths
  worksheet.columns.forEach((column) => {
    let maxLength = 10;
    column.eachCell({ includeEmpty: true }, (cell) => {
      const valStr = cell.value ? String(cell.value) : '';
      if (valStr.length > maxLength) {
        maxLength = valStr.length;
      }
    });
    column.width = Math.min(45, Math.max(12, maxLength + 3));
  });

  // 4. Output as Blob
  const buffer = await workbook.xlsx.writeBuffer();
  return new Blob([buffer], {
    type: 'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet',
  });
}

File: services/ocr_service/excel_generator.py
# excel_generator.py — OpenPyXL Excel Workbook (.xlsx) Generator with Multi-Sheet Support
"""
Generates genuine Microsoft Excel (.xlsx) workbooks with:
 - Royal Blue (#2563EB) styled header row
 - Zebra striping for improved legibility
 - Auto-adjusted column widths
 - Thin grid borders
 - Automatic numeric and currency type conversion
 - Multi-table & Multi-sheet support (e.g. Purchase, Sale, Stock sheets)
"""

import io
import re
from typing import List, Dict, Any, Optional
import openpyxl
from openpyxl.styles import Font, PatternFill, Alignment, Border, Side
from openpyxl.utils import get_column_letter


def format_worksheet(ws, headers: List[str], rows: List[List[str]]):
    ws.views.sheetView[0].showGridLines = True

    header_font = Font(name="Segoe UI", size=11, bold=True, color="FFFFFF")
    header_fill = PatternFill(start_color="2563EB", end_color="2563EB", fill_type="solid")
    header_align = Alignment(horizontal="center", vertical="center", wrap_text=True)

    data_font = Font(name="Segoe UI", size=10.5, color="1E293B")
    data_fill_even = PatternFill(start_color="FFFFFF", end_color="FFFFFF", fill_type="solid")
    data_fill_odd = PatternFill(start_color="F8FAFC", end_color="F8FAFC", fill_type="solid")

    thin_border = Border(
        left=Side(style='thin', color='E2E8F0'),
        right=Side(style='thin', color='E2E8F0'),
        top=Side(style='thin', color='E2E8F0'),
        bottom=Side(style='thin', color='E2E8F0')
    )
    header_border = Border(
        left=Side(style='thin', color='1D4ED8'),
        right=Side(style='thin', color='1D4ED8'),
        top=Side(style='thin', color='1D4ED8'),
        bottom=Side(style='medium', color='1E40AF')
    )

    # 1. Write Header Row
    if headers and any(h.strip() for h in headers):
        ws.append(headers)
        ws.row_dimensions[1].height = 28
        for col_idx in range(1, len(headers) + 1):
            cell = ws.cell(row=1, column=col_idx)
            cell.font = header_font
            cell.fill = header_fill
            cell.alignment = header_align
            cell.border = header_border

    # 2. Write Data Rows
    start_row = 2 if (headers and any(h.strip() for h in headers)) else 1
    for r_idx, row_data in enumerate(rows, start=start_row):
        typed_row = []
        for val in row_data:
            s_val = str(val or "").strip()
            # Clean currency / comma formatting for numeric check
            clean_s = s_val.replace("$", "").replace(",", "")
            if clean_s and (clean_s.replace(".", "", 1).replace("-", "", 1).isdigit()):
                try:
                    typed_row.append(float(clean_s) if "." in clean_s else int(clean_s))
                except ValueError:
                    typed_row.append(s_val)
            else:
                typed_row.append(s_val)

        ws.append(typed_row)
        ws.row_dimensions[r_idx].height = 22
        is_odd = (r_idx % 2 == 1)
        fill = data_fill_odd if is_odd else data_fill_even

        for col_idx in range(1, len(typed_row) + 1):
            cell = ws.cell(row=r_idx, column=col_idx)
            cell.font = data_font
            cell.fill = fill
            cell.border = thin_border

            if isinstance(cell.value, (int, float)):
                cell.alignment = Alignment(horizontal="right", vertical="center")
            else:
                cell.alignment = Alignment(horizontal="left", vertical="center", wrap_text=False)

    # 3. Auto-calculate Column Widths
    for col in ws.columns:
        max_len = 10
        col_letter = get_column_letter(col[0].column)
        for cell in col:
            if cell.value is not None:
                max_len = max(max_len, len(str(cell.value)))
        ws.column_dimensions[col_letter].width = min(45, max(12, max_len + 3))


def generate_xlsx(
    headers: Optional[List[str]] = None,
    rows: Optional[List[List[str]]] = None,
    sheet_name: str = "Table Data",
    tables: Optional[List[Dict[str, Any]]] = None
) -> bytes:
    """
    Generate a genuine formatted Microsoft Excel (.xlsx) workbook.
    If `tables` is provided (multi-table), creates separate sheets for each table.
    Otherwise, creates a single sheet with `headers` and `rows`.
    """
    wb = openpyxl.Workbook()

    if tables and len(tables) > 1:
        # Multi-sheet workbook
        wb.remove(wb.active) # remove default sheet
        for idx, tbl in enumerate(tables):
            name = tbl.get("name") or f"Table {idx+1}"
            safe_name = re.sub(r'[:\\/\?\*\[\]]', '', name)[:31]
            ws = wb.create_sheet(title=safe_name)
            format_worksheet(ws, tbl.get("headers", []), tbl.get("rows", []))
    else:
        # Single sheet workbook
        ws = wb.active
        tbl_headers = headers or (tables[0].get("headers", []) if tables else [])
        tbl_rows = rows or (tables[0].get("rows", []) if tables else [])
        title = sheet_name
        if tables and tables[0].get("name"):
            title = tables[0]["name"]
        safe_title = re.sub(r'[:\\/\?\*\[\]]', '', title)[:31]
        ws.title = safe_title
        format_worksheet(ws, tbl_headers, tbl_rows)

    out = io.BytesIO()
    wb.save(out)
    return out.getvalue()
